subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProvider.java

```diff
@@ -73,6 +76,15 @@ public BuildServiceProvider(
         this.isolatableFactory = isolatableFactory;
         this.internalServices = internalServices;
         this.listener = listener;
+        this.maxUsages = maxUsages;
+    }
+
+    public int getMaxUsages() {
+        return maxUsages;
+    }
+
+    public boolean isLockRequired() {
```

Comment:
❓ What's going on here?

Reply:
This is to allow the nagger to determine whether a service reference requires locking but is not part of the list of resources locked by the task. Not all service references require locks (for instance, locks are not created for services that do not impose maximum parallel usage).

Before, we required every service to be declared in the list of required services, so just checking that list in the nagger [was enough](https://github.com/gradle/gradle/blob/0dad3eab937f2b0f3bea0daaf42ef13c828c401e/subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProviderNagger.java#L46).  As things were implemented in this spike, that list holds only the list of service names explicitly declared as used (either using #usesService()  or using the name in the `@ServiceReference` annotation), but does not include the list of services injected manually. The actual list we care about is the list of locks, as checked [here](https://github.com/gradle/gradle/blob/0dad3eab937f2b0f3bea0daaf42ef13c828c401e/subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProviderNagger.java#L37).

Reply:
I see, thanks. For the final implementation it would be better not to conflate these things, i.e., service reference properties and changing how locks are checked, if possible.